### PR TITLE
Feature/issue 1262 back button for feedback sends to wrong screen

### PR DESCRIPTION
**Pull request by tomy-pg**

### CODE REVIEW

File: src/application/routing.ts [hunk 1]
```diff
@@ -115,11 +115,22 @@ const previousPathMatchesContactInformationRoute = (memoryHistory: MemoryHistory
     return pathMatchesRoute(previousPath, Routes.ContactInformation);
 };
 
+export const clearFeedbackPathsFromHistory = (memoryHistory: MemoryHistory): void => {
+    while (R.findLast(pathMatchesFeedbackRoute, memoryHistory.entries) && R.not(R.isEmpty(memoryHistory.entries))) {
```
Comment (tomy-pg) [inline src/application/routing.ts:115]: test to see if I can drop `R.not(R.isEmpty(memoryHistory.entries)`

File: src/components/feedback/contact_information_component.tsx [hunk 2]
```diff
@@ -54,14 +53,14 @@ export const ContactInformationComponent = ({
...
-        goBack(history);
+        goBack(memoryHistory);
         backFromContactInformation();
     };
 
     const onFinishPress = (): void => {
         finishFeedback(userInformation);
         sendFeedback(serviceId);
-        goToRouteWithParameter(Routes.ServiceDetail, serviceId, history)();
+        goBackToServiceDetailOnFeedbackSubmit(memoryHistory);
```
Comment (rasmus-storjohann-PG) [inline src/components/feedback/contact_information_component.tsx:63]: name should be popFeedbackElementsFromHistory

File: src/application/__tests__/routing.test.ts [hunk 3]
```diff
@@ -82,3 +86,89 @@ describe('the routePathWithParameter function', () => {
...
+    const serviceDetailPath = routePathDefinition(Routes.ServiceDetail);
+    const firstFeedbackPath = routePathDefinition(Routes.OtherFeedback);
+    const secondFeedbackPath = routePathDefinition(Routes.ContactInformation);
+
+    it('sends the user back to the service detail page already on the history stack', () => {
+        const initialPathEntries = [firstRegularPath, secondRegularPath, serviceDetailPath, firstFeedbackPath, secondFeedbackPath];
+        const indexOfLastPath = initialPathEntries.length - 1;
+        const history = createMemoryHistory({ initialEntries: initialPathEntries, initialIndex: indexOfLastPath});
+        goBackToServiceDetailOnFeedbackSubmit(history);
+        expect(history.location.pathname).toBe(serviceDetailPath);
```
Comment (rasmus-storjohann-PG) [inline src/application/__tests__/routing.test.ts:161]: Test that you have these `firstRegularPath, secondRegularPath, serviceDetailPath` all in order in the history

File: src/application/routing.ts [hunk 4]
```diff
@@ -93,16 +94,31 @@ export const routePathWithParameter = (route: Routes, parameter: string): string
...
+    clearFeedbackPathsFromHistory(memoryHistory);
+    const indexOfCurrentPathInHistoryStack = 0;
+    memoryHistory.go(indexOfCurrentPathInHistoryStack);
+};
+
+export const clearFeedbackPathsFromHistory = (memoryHistory: MemoryHistory): void => {
+    while (R.findLast(pathMatchesFeedbackRoute, memoryHistory.entries) && R.not(R.isEmpty(memoryHistory.entries))) {
+        memoryHistory.entries.pop();
+    }
+};
```
Comment (rasmus-storjohann-PG) [inline src/application/routing.ts:118]: Rename to popFeedbackPathsFromHistory

File: src/application/__tests__/routing.test.ts [hunk 5]
```diff
@@ -82,3 +86,89 @@ describe('the routePathWithParameter function', () => {
...
+        expect(history.location.pathname).toBe(learnPath);
+    });
+
+    it('Sends users to the "/feedback/:serviceId" path from "/contact-information/:serviceId"', () => {
+        const initialPathEntries = [servicesPath, serviceDetailPath, otherFeedbackPath, contactInformationPath];
+        const indexOfLastPath = initialPathEntries.length - 1;
+        const history = createMemoryHistory({ initialEntries: initialPathEntries, initialIndex: indexOfLastPath });
+        goBack(history);
+        expect(history.location.pathname).toBe(otherFeedbackPath);
+    });
```
Comment (rasmus-storjohann-PG) [inline src/application/__tests__/routing.test.ts:113]: I don't think this test is needed anymore? If you keep it, I think the name of the test needs to change.

File: src/application/__tests__/routing.test.ts [hunk 6]
```diff
@@ -82,3 +86,89 @@ describe('the routePathWithParameter function', () => {
...
+
+    it('Sends users to the "/service/:serviceId" path from "/feedback/:serviceId"', () => {
+        const initialPathEntries = [serviceDetailPath, otherFeedbackPath];
+        const history = createMemoryHistory({ initialEntries: initialPathEntries });
+        goBack(history);
+        expect(history.location.pathname).toBe(serviceDetailPath);
+    });
```
Comment (rasmus-storjohann-PG) [inline src/application/__tests__/routing.test.ts:120]: Same with this test, goBack() is not your function (right?) so both of these tests just show that it goes back one step, which is the same functionality in each test.

File: src/application/__tests__/routing.test.ts [hunk 7]
```diff
@@ -82,3 +86,89 @@ describe('the routePathWithParameter function', () => {
...
+
+    it('Sends users to the "/services/:topicId" from "/service/:serviceId"', () => {
+        const initialPathEntries = [servicesPath, serviceDetailPath];
+        const history = createMemoryHistory({ initialEntries: initialPathEntries });
+        goBack(history);
+        expect(history.location.pathname).toBe(servicesPath);
+    });
```
Comment (rasmus-storjohann-PG) [inline src/application/__tests__/routing.test.ts:127]: This test is also redundant I think.

File: src/application/routing.ts [hunk 8]
```diff
@@ -93,16 +94,31 @@ export const routePathWithParameter = (route: Routes, parameter: string): string
...
+
+const pathMatchesFeedbackRoute = (location: Location<LocationState>): boolean => (
+    pathMatchesAnyRoute(location.pathname, [Routes.OtherFeedback, Routes.ContactInformation])
```
Comment (rasmus-storjohann-PG) [inline src/application/routing.ts:121]: Editable service detail and remove service routes need to be in here too I think?

File: src/components/header_button/back_button_component.tsx [hunk 9]
```diff
@@ -3,15 +3,17 @@ import { I18nManager } from 'react-native';
 import { History } from 'history';
 import { Button, Icon } from 'native-base';
 import { goBack } from '../../application/routing';
+import { memoryHistory } from '../../application';
 
+// TO DO remove history prop.
```
Comment (rasmus-storjohann-PG) [inline src/components/header_button/back_button_component.tsx:8]: don't forget

**Review by rasmus-storjohann-PG — CHANGES_REQUESTED** [review]
Found some things after all, make sure you cover the three different feedback flows, remove service, other feedback and service feedback

File: src/application/routing.ts (hunk 8 above)
Comment (tomy-pg) [reply src/application/routing.ts:121]: These two screens are covered.
the `EditableServiceDetail` screen is equivalent to Routes.ServiceDetail. Its a state change that makes that screen editable, we don't change routes when this occurs. 
As well the `RemoveService` screen is equivalent to Routes.OtherFeedback.

Comment (tomy-pg) [reply src/application/routing.ts:121]: I will add a test to include the editableServiceDetail flow.

File: src/application/routing.ts (hunk 4 above)
Comment (tomy-pg) [reply src/application/routing.ts]: Fixed

File: src/application/__tests__/routing.test.ts (hunk 3 above)
Comment (tomy-pg) [reply src/application/__tests__/routing.test.ts:149]: Fixed

File: src/application/__tests__/routing.test.ts [hunk 10]
```diff
@@ -82,3 +86,81 @@ describe('the routePathWithParameter function', () => {
...
+
+    it('pops all feedback paths from the history stack when submitting feedback from the editable service detail screen', () => {
+        const initialPathEntries = [firstRegularPath, secondRegularPath, editableServiceDetailFeedbackPath, contactInformationFeedbackPath];
+        const history = createMemoryHistory({ initialEntries: initialPathEntries });
+        popFeedbackPathsFromHistory(history);
+        expect(history.entries.length).toBe(3);
+        expect(history.entries[0].pathname).toBe(firstRegularPath);
+        expect(history.entries[1].pathname).toBe(secondRegularPath);
+        expect(history.entries[2].pathname).toBe(serviceDetailPath);
```
Comment (rasmus-storjohann-PG) [inline src/application/__tests__/routing.test.ts:132]: Didn't expect this unless serviceDetailPath is the last element before editableServiceDetailFeedbackPath in the initialPathEntries

File: src/components/header_button/back_button_component.tsx (hunk 9 above)
Comment (rasmus-storjohann-PG) [reply src/components/header_button/back_button_component.tsx:7]: You can remove the comment now ;-)

**Review by rasmus-storjohann-PG — CHANGES_REQUESTED** [review]
Get Ian's help tomorrow to do a very careful QA pass on this, then you can merge it and push to staging.

**Review by ian-lin-PG — APPROVED** [review]

**Pull request merged**